lib/AST/ASTBuilder.cpp

```diff
@@ -536,4 +536,21 @@ clang::NullStmt *ASTBuilder::CreateNullStmt() {
   return new (ctx) clang::NullStmt(clang::SourceLocation());
 }
 
+clang::SwitchStmt *ASTBuilder::CreateSwitchStmt(clang::Expr *cond) {
```

Comment:
Are there no methods in `clang::Sema` to create these? Just to get some extra bit of semantic checking.

Reply:
`clang::Sema::CheckSwitchCondition` works fine, but `clang::Sema::ActOnCaseExpr` only works on `case`s that are already in a switch statement, I believe

Reply:
Yeah, that's not surprising. Sometimes we can use sema, sometimes we can't. Hmm. Would it be practical to create the whole `switch` in one go with all the cases? So that we'd have something like `CreateSwitchStmt(clang::Expt *cond, std::vector<std::pair<clang::Expr*, clang::Stmt*>> &cases)` where the pair contains the case ordinal and case code.

Reply:
Hrm it's not that easy, the statements under a `case` are not actually substatements of the `case` itself, the structure is more flat, with all the cases and their statements belonging to the compound statement of the switch itself